### PR TITLE
Refactored indexing algorithm and DefinitionsDatabase

### PR DESCRIPTION
**Pull request by m-hilgendorf**
- Added support for visibility modifiers in the definitions database
- Altered algorithm to flatten IL-IR into definitions and their visibility
- Split indexing functions into helpers to make it easier to test
- Added unit tests to check each step of the indexing process.

### CODE REVIEW

File: compiler/stz-defs-db-serializer.stanza [hunk 1]
```diff
@@ -23,31 +24,61 @@ public public defenum SrcDefinitionSource :
...
+public defstruct PackageDefinition : 
+  name:Symbol
+  file-name:String
+  imports:Tuple<PackageImport>
+
+public defstruct PackageImport : 
+  name:Symbol
+  prefixes:Tuple<PackageImportPrefix>
+
+public defstruct PackageImportPrefix :
```
Comment (CuppoJava) [inline compiler/stz-defs-db-serializer.stanza:50]: [P] Thanks this is necessary, but it seems to be unused currently.
What is supposed to create this structure? I also recommend changing
the names and types to more closely match their original names in
`il-ir.stanza`. See the `IPrefix` struct in `il-ir.stanza`.

So:
```
  sources => names
  Tuple<Symbol> => False|Tuple<Symbol>
  prefix:Symbol => prefix:String
```

The `False` case for `names` is used when the user types:

  prefix => pre-

Which indicates to use a single prefix for absolutely everything. See
the `IPrefix` type in `il-ir.stanza`.

After you change `Tuple<Symbol>` to `False|Tuple<Symbol>`, you'll have
to copy over `read-opt` and `write-opt` from `stz-pkg.stanza` for the
serialization. See the usages of `opt<FileInfo>(...)` for examples on
how they are used.

public defstruct IPrefix :
  names: False|Tuple<Symbol>
  prefix: String
with:
  printer => true

File: compiler/stz-defs-db.stanza [hunk 2]
```diff
@@ -19,27 +19,38 @@ defpackage stz/defs-db :
 ;============================================================
 ;=================== Driver Arguments =======================
 ;============================================================
-
 public defstruct DefsDbInput :
   proj-files: Tuple<String>
   platform: Symbol|False
```
Comment (CuppoJava) [inline compiler/stz-defs-db.stanza]: Don't allow this platform to be false. Fill it with OUTPUT-PLATFORM when this is created.

File: compiler/stz-defs-db.stanza [hunk 3]
```diff
@@ -19,27 +19,38 @@ defpackage stz/defs-db :
...
+
+; Helper function to compile the input data into IL-IR.
+protected defn compile-input (input:DefsDbInput) -> [NameMap, Seqable<IPackage|Pkg>] :
   ;Compute platform
   read-config-file()
-  val db-platform = match(platform(input)) :
-    (s:Symbol) : s
-    (f:False) : OUTPUT-PLATFORM
-    
+  val db-platform =
```
Comment (CuppoJava) [inline compiler/stz-defs-db.stanza]: Move this logic out of here. See the comment beside `platform:Symbol|False`.

File: compiler/stz-defs-db.stanza [hunk 4]
```diff
@@ -53,111 +64,128 @@ public defn defs-db (input:DefsDbInput, filename:String) :
...
+; Expected Behavior : 
+; - When encountering an `IVisibility` node, traverse its child using that node's visibility
+; - When encountering an `IBegin`, traverse its children with the current visibility.
+; - The initial visibility is `Private`
+; - If an expression is not `Indexable`, it is skipped
+; - If an `Indexable` does not have an `info`, it is skipped
+; - If an `Indexable`'s name cannot be found, it is skipped 
+;
+protected defn index-expressions (exps:Seqable<IExp>, nm:NameMap) -> Seq<Indexed> :
+  var print? = false
```
Comment (CuppoJava) [inline compiler/stz-defs-db.stanza]: Debug code.

File: compiler/stz-defs-db.stanza [hunk 5]
```diff
@@ -53,111 +64,128 @@ public defn defs-db (input:DefsDbInput, filename:String) :
...
+  generate : 
+    defn loop (e:IExp, current-visibility:Visibility) :
+      match(e) : 
+        ; Case 1: We encounter a visibility node. Walk its
```
Comment (CuppoJava) [inline compiler/stz-defs-db.stanza:119]: I like the structure of the comments here. It makes both the intention and implementation of the algorithm very clear.

File: compiler/stz-defs-db.stanza [hunk 6]
```diff
@@ -53,111 +64,128 @@ public defn defs-db (input:DefsDbInput, filename:String) :
...
+; - When encountering an `IVisibility` node, traverse its child using that node's visibility
+; - When encountering an `IBegin`, traverse its children with the current visibility.
+; - The initial visibility is `Private`
+; - If an expression is not `Indexable`, it is skipped
+; - If an `Indexable` does not have an `info`, it is skipped
+; - If an `Indexable`'s name cannot be found, it is skipped 
+;
+protected defn index-expressions (exps:Seqable<IExp>, nm:NameMap) -> Seq<Indexed> :
+  var print? = false
+  generate :
```
Comment (CuppoJava) [inline compiler/stz-defs-db.stanza]: `generate<Indexed>` for clarity.

**Review by CuppoJava — CHANGES_REQUESTED** [review]
Fantastic work. Implementation is both more powerful and cleaner than before.

File: compiler/stz-defs-db.stanza (hunk 2 above)
Comment (m-hilgendorf) [reply compiler/stz-defs-db.stanza]: This comes directly from the CLI command, so I added it to `stz-main.stanza`. It turns out that the previous implementation always gave `false` for `platform` and it's not immediately useful to support cross-compilation for indexing commands, so I hardcoded it to OUTPUT-PLATFORM.

File: compiler/stz-defs-db.stanza (hunk 4 above)
Comment (m-hilgendorf) [reply compiler/stz-defs-db.stanza]: Removed! Thank you

File: compiler/stz-defs-db.stanza (hunk 6 above)
Comment (m-hilgendorf) [reply compiler/stz-defs-db.stanza]: Fixed!

File: compiler/stz-defs-db-serializer.stanza (hunk 1 above)
Comment (m-hilgendorf) [reply compiler/stz-defs-db-serializer.stanza:50]: I split out the construction of `PackageDefinition` objects into a helper called `index-definitions` that now appropriately handles the extraction of imports and prefixes. 

Thank you for the pointer on `read-opt` and `write-opt`!

File: compiler/stz-defs-db.stanza (hunk 3 above)
Comment (m-hilgendorf) [reply compiler/stz-defs-db.stanza]: Removed!

**Pull request merged**